[concurrency] type-checking call-sites that involve an async throws / rethrows callee

kavon opened the pull request:
A combo fix for problems when type-checking a call-site to an async and throwing function. These issues are quite similar and deal with the same part of the compiler, so I'm doing one PR for it (for now).

## Part 1

Suppose we have,

```
func f (fn: () async throws -> Int) async rethrows -> Int {
  return await try fn()
}
```

If a call to `f` appears in a synchronous context, the type-checker was not complaining as it should. The root cause is that while checking this `rethrows` callee, we forgot to carry along metadata about the call-site being async.

## Part 2

Suppose we have,

```
 func throwingAndAsync() async throws -> Int { return 0 }
```

If we make a call to `throwingAndAsync` under a try context that is optional (`try?`) or forced (`try!`), the call-site is not correctly recognized as being `async`.



### TODOs

- [x] Fix for Part 1 (rdar://70751405)
  - [x] Regression tests.
- [x] Fix for Part 2 (rdar://70813762)
  - [x] Regression tests.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves rdar://70751405
Resolves rdar://70813762

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

kavon committed:
Consider 'async' during classification of call-sites to rethrows functions.

kavon committed:
add regression test for rdar://70751405 and rdar://70813762

DougGregor commented on lib/Sema/TypeCheckEffects.cpp:
```diff
@@ -461,7 +461,8 @@ class ApplyClassifier {
 
       // Use the most significant result from the arguments.
-      Classification result;
+      Classification result = isAsync ? Classification::forAsync() 
```
Hah, nice find.

kavon committed:
add a dumper for effect Classification for debugging

kavon committed:
add isAsync param to forRethrowingOnly for consistency with forThrow

kavon committed:
fix bug when merging Classifications that was clobbering IsAsync

kavon committed:
fix missing propagation of async-ness through try? and try!

kavon committed:
update test now that rdar://70751405 is fixed

kavon commented on test/Concurrency/async_throwing.swift:
```diff
@@ -0,0 +1,138 @@
+  let _ = await invokeAuto(await try? throwingTask())
+
+  // FIXME: we currently emit a less-helpful error message here.
+  // it would be better if we said "call is 'async' in an autoclosure argument that is not marked with 'await'"
```
Should I also try to fix this error message in this PR?

ktoso replied:
It would be nice to get this in since it cleans up many callsites of tasks, and work on follow ups separately perhaps?

kavon replied:
It ended up being simple enough to add to this PR.

ktoso commented on test/Concurrency/async_cancellation.swift:
```diff
@@ -53,14 +53,13 @@ func test_cancellation_loop() async -> Int {
 
 func test_cancellation_withDeadline_in() async throws -> Int {
-  /* missing await */ Task.withDeadline(in: .seconds(5), operation: { // FIXME: rdar://70751405 async rethrows functions are not detected as async
+  await Task.withDeadline(in: .seconds(5), operation: {
```
Wohoo 👍

ktoso commented:
Mind if I kick off tests here?

@swift-ci please smoke test

kavon committed:
fix cases where error message was not specific when missing an 'await' in autoclosure argument.

when the call in an autoclosure is async and throwing, if the call is covered by a try! or try?,
the existing check that used Context::isAutoClosure was too pessimistic, saying it is not an
autoclosure because errors are handled within it. For the error message about a missing await,
we don't care about whether the throwing-ness was handled.

DougGregor commented:
@swift-ci smoke test
